src/Compilers/Core/Portable/DiagnosticAnalyzer/DiagnosticAnalysisContext.cs

```diff
@@ -1166,11 +1171,17 @@ public struct OperationAnalysisContext
         /// </summary>
         public CancellationToken CancellationToken => _cancellationToken;
 
+        [Obsolete("Prefer the constructor that takes SemanticModel instead of Compilation")]
         public OperationAnalysisContext(IOperation operation, ISymbol containingSymbol, Compilation compilation, AnalyzerOptions options, Action<Diagnostic> reportDiagnostic, Func<Diagnostic, bool> isSupportedDiagnostic, CancellationToken cancellationToken)
+            : this(operation, containingSymbol, compilation.GetSemanticModel(operation.Syntax.SyntaxTree), options, reportDiagnostic, isSupportedDiagnostic, cancellationToken)
```

Comment:
>compilation.GetSemanticModel(operation.Syntax.SyntaxTree) [](start = 48, length = 57)

I don't think this is the right thing to do. Why would we want to allocate new SemanticModel instance every time this constructor is used? I think it is better to keep compilation as a separate field and SemanticModel as a separate field that can be null, depending on whether it was given at construction.

Reply:
BTW, given current implementation for IOperation, the SemanticModel and it's Compilation can always be obtained form an IOperation instance. Why force people passing them explicitly at all?

---
In reply to: [182534943](https://github.com/dotnet/roslyn/pull/26187#discussion_r182534943) [](ancestors = 182534943)

Reply:
>I don't think this is the right thing to do.
It is not nice but the ctor was deprecated and all usages removed. Please suggest if there is a nicer way.

Reply:
> SemanticModel and it's Compilation can always be obtained form an IOperation instance.

How?

Reply:
Given that this is a public constructor, we cannot really assume or direct cast `IOperation` to [Operation](http://source.roslyn.io/#Microsoft.CodeAnalysis/Operations/Operation.cs,c6a610c4883cad18) to fetch the underlying `SemanticModel`. I think we should do the following: `(operation as Operation)?.SemanticModel ?? compilation.GetSemanticModel(operation.Syntax.SyntaxTree)`

Reply:
> compilation.GetSemanticModel(operation.Syntax.SyntaxTree)

We shouldn't allocate new SemanticModel this is going to lead to perf issues. 

---
In reply to: [182549821](https://github.com/dotnet/roslyn/pull/26187#discussion_r182549821) [](ancestors = 182549821)

Reply:
> Given that this is a public constructor, we cannot really assume or direct cast IOperation to Operation to fetch the underlying SemanticModel. 

Then we simply shouldn't return any SemanticModel for an IOperation that was not created by a SemanticModel

---
In reply to: [182550732](https://github.com/dotnet/roslyn/pull/26187#discussion_r182550732) [](ancestors = 182550732,182549821)

Reply:
> Then we simply shouldn't return any SemanticModel for an IOperation that was not created by a SemanticModel

That is correct, we should probably just use `(operation as Operation)?.SemanticModel`. Does this mean that we should make the newly added constructor overload public for third party implementors of `IOperation`? Probably we can do so once we have one such client requesting it.

Reply:
I'd say, let's wait for a real use case before adding a new public constructor. 

---
In reply to: [182564629](https://github.com/dotnet/roslyn/pull/26187#discussion_r182564629) [](ancestors = 182564629)